File: subarraySum.cpp
 int subarraySum(vector<int>& nums, int k) {
        
        int n = nums.size();
        
        if(n == 0)
            return 0;
        
        if(n == 1)
        {
            if(nums[0] == k)
                return 1;

            return 0;
        }

        unordered_map<int, int> hashPrefixSum;
        
        hashPrefixSum[0] = 1;

        int subArray = 0;

        int sum = 0;
        
        for(int i = 0; i < n; i++)
        {
            sum += nums[i];
            
            if(hashPrefixSum.find(sum - k) != hashPrefixSum.end())
                    subArray += hashPrefixSum[sum - k];
            
            hashPrefixSum[sum]++;
        }
        
        return subArray;
       
    }

File: progress_sheet/Subarraysum.py
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:

        prefix_sum = defaultdict(int)
        pre_sum = 0
        subarrays = 0
        size = len(nums)

        # count subarray with sum k and 
        # count subarray by chopping off the current part of subarray having total sum = prefix sum - k
        for i in range(size):

            pre_sum += nums[i]

            if pre_sum == k:
                subarrays += 1

            subarrays += prefix_sum[pre_sum - k]
            prefix_sum[pre_sum] += 1

        return subarrays  
